src/components/molecules/timeline/timeline.tsx

```diff
@@ -1,17 +1,33 @@
 import { Fragment } from 'react';
 import './timeline.scss';
 
-export type TimelineEvent = {
+type TimelineEventBasic = {
   name: string,
   date: string,
-  link?: {
+};
+
+type TimelineEventSimple = TimelineEventBasic & {
+  link?: undefined,
+  eventDescription?: undefined,
+};
+
+type TimelineEventWithLink = TimelineEventBasic & {
+  link: {
     text: string,
-    url: string,
-  }
+    url?: string,
+  },
+  eventDescription?: undefined,
 };
 
+type TimelineEventWithDescription = TimelineEventBasic & {
+  link?: undefined,
+  eventDescription: string,
+};
+
+export type TimelineEvent = TimelineEventSimple | TimelineEventWithLink | TimelineEventWithDescription;
```

Comment:
I had initially approached it as follows but couldn't get it to work:

```
type TimelineEventBasic = {
  name: string,
  date: string,
};

type TimelineEventWithLink = TimelineEventBasic & {
  link: {
    text: string,
    url?: string,
  },
};

type TimelineEventWithDescription = TimelineEventBasic & {
  eventDescription: string,
};

export type TimelineEvent = TimelineEventBasic | TimelineEventWithLink | TimelineEventWithDescription;
```

Reply:
[Because](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#working-with-union-types):
> TypeScript will only allow an operation if it is valid for every member of the union. For example, if you have the union string | number, you can’t use methods that are only available on string:

The way around it is with typeGuard functions:
```
const isTimelineEventWithLink = (event: TimelineEvent): event is TimelineEventWithLink => (event as TimelineEventWithLink).link === undefined;
const isTimelineEventWithDescription = (event: TimelineEvent): event is TimelineEventWithDescription => (event as TimelineEventWithDescription).eventDescription === undefined;
const isTimelineEventBasic = (event: TimelineEvent): event is TimelineEventBasic => !isTimelineEventWithLink(event) && !isTimelineEventWithDescription(event);
```

then you could change the code like so:

```
diff --git a/src/components/molecules/timeline/timeline.tsx b/src/components/molecules/timeline/timeline.tsx
index 046c1eb..734adae 100644
--- a/src/components/molecules/timeline/timeline.tsx
+++ b/src/components/molecules/timeline/timeline.tsx
@@ -6,25 +6,22 @@ type TimelineEventBasic = {
   date: string,
 };
 
-type TimelineEventSimple = TimelineEventBasic & {
-  link?: undefined,
-  eventDescription?: undefined,
-};
-
 type TimelineEventWithLink = TimelineEventBasic & {
   link: {
     text: string,
     url?: string,
   },
-  eventDescription?: undefined,
 };
 
 type TimelineEventWithDescription = TimelineEventBasic & {
-  link?: undefined,
   eventDescription: string,
 };
 
-export type TimelineEvent = TimelineEventSimple | TimelineEventWithLink | TimelineEventWithDescription;
+export type TimelineEvent = TimelineEventBasic | TimelineEventWithLink | TimelineEventWithDescription;
+
+const isTimelineEventWithLink = (event: TimelineEvent): event is TimelineEventWithLink => (event as TimelineEventWithLink).link === undefined;
+const isTimelineEventWithDescription = (event: TimelineEvent): event is TimelineEventWithDescription => (event as TimelineEventWithDescription).eventDescription === undefined;
+// const isTimelineEventBasic = (event: TimelineEvent): event is TimelineEventBasic => !isTimelineEventWithLink(event) && !isTimelineEventWithDescription(event);
 
 type TimelineProps = {
   events: Array<TimelineEvent>,
@@ -42,8 +39,8 @@ export const Timeline = ({ events, listDescription }: TimelineProps): JSX.Elemen
             <dt className="review-timeline__event">{entry.name}</dt>
             <dd className="review-timeline__date">
               {formatDate(entry.date)}
-              {entry.link && <a className="review-timeline__link" href={entry.link.url} aria-label={entry.name}>{entry.link.text}</a>}
-              {entry.eventDescription && <span className="review-timeline__description"> {entry.eventDescription}</span>}
+              {isTimelineEventWithLink(entry) && <a className="review-timeline__link" href={entry.link.url} aria-label={entry.name}>{entry.link.text}</a>}
+              {isTimelineEventWithDescription(entry) && <span className="review-timeline__description"> {entry.eventDescription}</span>}
             </dd>
           </Fragment>
         ))
```

I think this is the cleaner way to express your original types, but at the cost of a few extra functions for type guarding.